pkg/lib/cockpit.d.ts

```diff
@@ -93,6 +93,49 @@ declare module 'cockpit' {
     function channel(options: ChannelOptions & { binary?: false; }): Channel<string>;
     function channel(options: ChannelOptions & { binary: true; }): Channel<Uint8Array>;
 
+    /* === cockpit.spawn ============================= */
+
+    interface DeferredPromise<T> extends Promise<T> {
+        /* jQuery Promise compatibility */
+        done(callback: (data: T) => void): DeferredPromise<T>
+        fail(callback: (exc: Error) => void): DeferredPromise<T>
+        always(callback: () => void): DeferredPromise<T>
+        progress(callback: (message: T, cancel: () => void) => void): DeferredPromise<T>
+    }
+
+    interface Deferred<T> {
+        resolve(): Deferred<T>;
+        reject(): Deferred<T>;
+        notify(): Deferred<T>;
+        promise: DeferredPromise<T>
+    }
+
+    interface Spawn<T> extends DeferredPromise<T> {
+        input(message: T, stream?: boolean): DeferredPromise<T>;
+        stream(callback: (data: T) => void): DeferredPromise<T>;
+        close(): void;
+    }
+
+    interface SpawnOptions {
+        binary?: boolean,
+        directory?: string;
+        err?: "out" | "ignore" | "message" | "pty";
+        environ?: string[];
+        pty?: boolean;
+        batch?: number;
+        latency?: number
+        superuser?: string;
+    }
+
+    function spawn(
+        args: string[],
+        options?: SpawnOptions & { binary?: false }
+    ): Spawn<string>
+    function spawn(
+        args: string[],
+        options?: SpawnOptions & { binary?: true }
```

Comment:
You really want `options: SpawnOptions & { binary: true }` here.  `binary?: true` can also match the case where `binary` is not specified and `options?:` can match if `options` is not specified, which means that the overload possibilities aren't disjoint.  I'm not sure what happens in that case, and I guess the compiler probably prefers the first match, but let's tighten this up...

Reply:
Sure! Need to re-read this a few times to fully understand it.